### PR TITLE
feat(manual_compaction): meta server support starting compaction

### PR DESCRIPTION
**Pull request by hycdong**
As apache/incubator-pegasus#850 shows, meta server should know the status of manual compaction.
This pull request adds a new meta server interface to start manual compaction and adds related unit test.

### CODE REVIEW

File: src/meta/test/meta_mauanl_compaction_test.cpp [hunk 1]
```diff
@@ -87,6 +140,40 @@ class meta_app_compaction_test : public meta_test_base
...
+    } tests[] = {{"app_not_exist", "false", false, -1, 0, ERR_APP_NOT_EXIST, "skip"},
+                 {APP_NAME, "true", false, -1, 0, ERR_OPERATION_DISABLED, "skip"},
+                 {APP_NAME, "false", false, -5, 0, ERR_INVALID_PARAMETERS, "skip"},
+                 {APP_NAME, "false", false, -1, -1, ERR_INVALID_PARAMETERS, "skip"},
+                 {APP_NAME, "false", false, -1, 0, ERR_OK, "skip"},
+                 {APP_NAME, "false", true, -1, 0, ERR_OK, "force"},
+                 {APP_NAME, "false", false, 1, 0, ERR_OK, "skip"},
+                 {APP_NAME, "false", true, -1, 1, ERR_OK, "force"}};
+
+    for (auto test : tests) {
```
Comment (levy5307) [inline src/meta/test/meta_mauanl_compaction_test.cpp:163]: for (const auto &test : tests) {

File: src/meta/server_state.cpp [hunk 2]
```diff
@@ -2943,6 +2944,118 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
...
+            ec == ERR_OK, "update app_info to remote storage failed with err = %s", ec.to_string());
+
+        zauto_write_lock l(_lock);
+        std::shared_ptr<app_state> app = get_app(app_name);
+        std::string old_envs = dsn::utils::kv_map_to_string(app->envs, ',', '=');
+        for (int idx = 0; idx < keys.size(); idx++) {
+            app->envs[keys[idx]] = values[idx];
+        }
+        std::string new_envs = dsn::utils::kv_map_to_string(app->envs, ',', '=');
+        ddebug("update manual compaction envs succeed: old_envs = {%s}, new_envs = {%s}",
```
Comment (levy5307) [inline src/meta/server_state.cpp:3043]: ddebug_f

File: src/meta/server_state.cpp [hunk 3]
```diff
@@ -2943,6 +2944,118 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
...
+    }
+    keys.emplace_back(replica_envs::MANUAL_COMPACT_ONCE_TRIGGER_TIME);
+    values.emplace_back(std::to_string(trigger_time));
+
+    // update new app_envs
+    for (auto idx = 0; idx < keys.size(); idx++) {
+        ainfo.envs[keys[idx]] = values[idx];
+    }
+    do_update_app_info(app_path, ainfo, [this, app_name, keys, values, rpc](error_code ec) {
+        dassert(
```
Comment (levy5307) [inline src/meta/server_state.cpp:3033]: dassert_f

File: src/meta/server_state.cpp [hunk 4]
```diff
@@ -2943,6 +2944,117 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
     return valid;
 }
 
+void server_state::on_start_manual_compact(start_manual_compact_rpc rpc)
```
Comment (levy5307) [inline src/meta/server_state.cpp:2947]: The function is too long

Comment (hycdong) [reply src/meta/server_state.cpp:2947]: I agree with you, this function will do following things:
1. check app validation
2. parse request parameter into manual compaction app envs
3. update app envs
4. reset local status

What's your suggestion about how to make think function shorter?

Comment (levy5307) [reply src/meta/server_state.cpp:2947]: split this function into four small functions, according to your comment above

File: src/meta/server_state.cpp [hunk 5]
```diff
@@ -2943,6 +2944,117 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
...
+    values.emplace_back(bottommost);
+
+    int64_t trigger_time = dsn_now_s();
+    if (request.__isset.trigger_time) {
+        trigger_time = request.trigger_time;
+    }
+    keys.emplace_back(replica_envs::MANUAL_COMPACT_ONCE_TRIGGER_TIME);
+    values.emplace_back(std::to_string(trigger_time));
+
+    // update new app_envs
```
Comment (levy5307) [inline src/meta/server_state.cpp:3028]: Good annotation is used for explaining why did you do like this, not what to do here.

File: src/meta/server_state.cpp (hunk 4 above)
Comment (hycdong) [reply src/meta/server_state.cpp:2947]: I add a new function `update_compaction_envs_on_remote_storage` to implement the step2 and step3, remain step1 and step4 in original `on_start_manual_compact` function.

File: src/meta/server_state.cpp (hunk 5 above)
Comment (hycdong) [reply src/meta/server_state.cpp]: I sometimes agree with you. In this case, the annotation is used as a division. But I also update it in my new commit.

**Review by Smityz — APPROVED** [review]

File: src/meta/server_state.cpp [hunk 6]
```diff
@@ -2943,6 +2944,132 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
...
 
+void server_state::on_start_manual_compact(start_manual_compact_rpc rpc)
+{
+    const std::string &app_name = rpc.request().app_name;
+    auto &response = rpc.response();
+
+    std::map<std::string, std::string> envs;
+    {
+        zauto_read_lock l(_lock);
+        std::shared_ptr<app_state> app = get_app(app_name);
```
Comment (levy5307) [inline src/meta/server_state.cpp:2955]: auto app = get_app(app_name);

File: src/meta/server_state.cpp [hunk 7]
```diff
@@ -2943,6 +2944,132 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
...
+    }
+
+    if (!update_compaction_envs_on_remote_storage(rpc)) {
+        return;
+    }
+
+    // update local manual compaction status
+    {
+        zauto_write_lock l(_lock);
+        std::shared_ptr<app_state> app = get_app(app_name);
```
Comment (levy5307) [inline src/meta/server_state.cpp:2983]: same

File: src/meta/server_state.cpp [hunk 8]
```diff
@@ -2943,6 +2944,132 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
...
+    }
+}
+
+bool server_state::update_compaction_envs_on_remote_storage(start_manual_compact_rpc rpc)
+{
+    const auto &request = rpc.request();
+    const std::string &app_name = request.app_name;
+    auto &response = rpc.response();
+
+    // parse manual compaction envs from request
```
Comment (levy5307) [inline src/meta/server_state.cpp:2994]: Use L2994~3039 to implement a function named `parse_manual_compaction_envs`

**Review by levy5307 — APPROVED** [review]

**Review by Smityz — APPROVED** [review]

**Pull request merged**